pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/UGICache.java

```diff
@@ -0,0 +1,318 @@
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.security.UserGroupInformation;
+
+/**
+ * Stores UserGroupInformation instances for each active session. The UGIs are cleaned up if they
+ * have not been accessed for 15 minutes.
+ * <p>
+ * The motivation for caching is that destroying UGIs is slow. The alternative, creating and
+ * destroying a UGI per-request, is wasteful.
+ */
+public class UGICache {
+
+    private static final Log LOG = LogFactory.getLog(UGICache.class);
+    private Map<SessionId, Entry> cache = new ConcurrentHashMap<>();
+    @SuppressWarnings("unchecked")
+    // There is a separate DelayQueue for each segment (also being used for locking)
+    private final Map<Integer, DelayQueue<Entry>> queueMap = new HashMap<>();
+    private final UGIProvider ugiProvider;
+    private Ticker ticker;
+    private final static long UGI_CACHE_EXPIRY = 15 * 60 * 1000L; // 15 Minutes
+
+    /**
+     * Create a UGICache with the given {@link UGIProvider}. Intended for use by tests which need
+     * to substitute a mock UGIProvider.
+     */
+    UGICache(UGIProvider provider, Ticker ticker) {
+        this.ticker = ticker;
+        this.ugiProvider = provider;
+    }
+
+    /**
+     * Create a UGICache. Automatically creates a {@link UGIProvider} that this cache will use to
+     * create and destroy UserGroupInformation instances.
+     */
+    public UGICache() {
+        this(new UGIProvider(), Ticker.systemTicker());
+    }
+
+    /**
+     * Create new proxy UGI if not found in cache and increment reference count
+     */
+    public UserGroupInformation getUserGroupInformation(SessionId session) throws IOException {
+        Integer segmentId = session.getSegmentId();
+        String user = session.getUser();
+        DelayQueue<Entry> delayQueue = getExpirationQueue(segmentId);
+        synchronized (delayQueue) {
+            // Use the opportunity to cleanup any expired entries
+            cleanup(session);
+            Entry entry = cache.get(session);
+            if (entry == null) {
+                LOG.info(session.toString() + " Creating proxy user = " + user);
+                entry = new Entry(ticker, ugiProvider.createProxyUGI(user));
+                delayQueue.offer(entry);
+                cache.put(session, entry);
+            }
+            entry.acquireReference();
+            return entry.getUGI();
+        }
+    }
+
+    /**
+     * Decrement reference count for the given session's UGI. Resets the time at which the UGI will
+     * expire to 15 minutes in the future.
+     *
+     * @param session                  the session for which we want to release the UGI.
+     * @param cleanImmediatelyIfNoRefs if true, destroys the UGI for the given session (only if it is
+     *                                 now unreferenced).
+     */
+    public void release(SessionId session, boolean cleanImmediatelyIfNoRefs) {
```

Comment:
I'd like the parameter names to allow someone reading this code to consider the cache in isolation from its callers, and from that perspective, `cleanImmediatelyIfNoRefs` or some variant seems like a good choice. The cache can't know or control whether it's the "lastCall" to `release`.

What if, instead of using a boolean parameter to change the behavior of this function, we had a separate method `destroyUnreferencedUGIs` that clients could call to clean things up? Then that method could be called in the `SecurityServletFilter` and the concept of "lastCall" could be represented there.

Reply:
hmm ... not sure that would be better. We have dilemma here -- cache should not know about caller and caller should not care about internal cache implementation, such as that cleaning is required at all, etc. Here're we're trying to cheat and give cache a bit more context (such as that this is our last call) and let it decide to clean immediately instead of waiting for it to timeout.

Either way we slice it it is not elegant. The proper way to separate concerns is perhaps registering a listener into a session that listens to close() events and does this `destroyUnreferencedUGIs` call and having a client calling session.close() when it is last call. This way client only closes session, session only calls a listener and the listener, which is part of the cache does the cleaning.

I do realize, however, this is a total overkill for the problem at hand. Whatever we call this doesn;t matter that much, the discussion itself is more interesting :)